daemon/container_unix.go

```diff
@@ -914,7 +916,7 @@ func (daemon *Daemon) allocateNetwork(container *Container) error {
 	}
 
 	for n := range container.NetworkSettings.Networks {
-		if err := daemon.connectToNetwork(container, n, updateSettings); err != nil {
+		if err := daemon.connectToNetwork(container, n, updateSettings, true); err != nil {
```

Comment:
this wont work for the container restart case. since all networks are treated equal during the restart case, it will not be possible to differentiate which network needs the port-mapping.
Hence, i would recommend, using the `hostConfig.NetworkMode` and compare that with the network being connected to before deciding to do the port-mapping.

That will also address @cpuguy83 comment on passing the flag.